lib/CrossTU/CrossTranslationUnit.cpp

```diff
@@ -338,33 +432,63 @@ llvm::Expected<ASTUnit *> CrossTranslationUnitContext::loadExternalAST(
+    StringRef ASTSource = It->second;
+    auto ASTCacheEntry = FileASTUnitMap.find(ASTSource);
     if (ASTCacheEntry == FileASTUnitMap.end()) {
-      IntrusiveRefCntPtr<DiagnosticOptions> DiagOpts = new DiagnosticOptions();
-      TextDiagnosticPrinter *DiagClient =
-          new TextDiagnosticPrinter(llvm::errs(), &*DiagOpts);
-      IntrusiveRefCntPtr<DiagnosticIDs> DiagID(new DiagnosticIDs());
-      IntrusiveRefCntPtr<DiagnosticsEngine> Diags(
-          new DiagnosticsEngine(DiagID, &*DiagOpts, DiagClient));
-
-      std::unique_ptr<ASTUnit> LoadedUnit(ASTUnit::LoadFromASTFile(
-          ASTFileName, CI.getPCHContainerOperations()->getRawReader(),
-          ASTUnit::LoadEverything, Diags, CI.getFileSystemOpts()));
-      Unit = LoadedUnit.get();
-      FileASTUnitMap[ASTFileName] = std::move(LoadedUnit);
+      if (!OnDemandParsingDatabase) {
+        // If no \p OnDemandParsingDatabase is given, try to load from AST dump
+        // file, as on-demand parsing is disabled.
+        IntrusiveRefCntPtr<DiagnosticOptions> DiagOpts =
+            new DiagnosticOptions();
+        TextDiagnosticPrinter *DiagClient =
+            new TextDiagnosticPrinter(llvm::errs(), &*DiagOpts);
+        IntrusiveRefCntPtr<DiagnosticIDs> DiagID(new DiagnosticIDs());
+        IntrusiveRefCntPtr<DiagnosticsEngine> Diags(
+            new DiagnosticsEngine(DiagID, &*DiagOpts, DiagClient));
+
+        std::unique_ptr<ASTUnit> LoadedUnit(ASTUnit::LoadFromASTFile(
+            ASTSource, CI.getPCHContainerOperations()->getRawReader(),
+            ASTUnit::LoadEverything, Diags, CI.getFileSystemOpts()));
+        Unit = LoadedUnit.get();
+
+        // Cache the resulting ASTUnit.
+        if (Unit)
+          FileASTUnitMap[ASTSource] = std::move(LoadedUnit);
+      } else {
+
+        // Lazily initialize the compilation database.
+        if (!CompileCommands) {
+          std::string LoadError;
+          CompileCommands = tooling::JSONCompilationDatabase::loadFromFile(
+              OnDemandParsingDatabase.getValue(), LoadError,
+              tooling::JSONCommandLineSyntax::AutoDetect);
+          if (!CompileCommands)
+            return llvm::make_error<IndexError>(
+                index_error_code::failed_to_get_external_ast);
+        }
+
+        // Try loading on-demand.
+        std::unique_ptr<ASTUnit> LoadedUnit = loadASTOnDemand(ASTSource);
+        Unit = LoadedUnit.get();
+
+        // Cache the sulting ASTUnit.
+        if (Unit)
+          FileASTUnitMap[ASTSource] = std::move(LoadedUnit);
+      }
       ++NumASTLoaded;
       if (DisplayCTUProgress) {
-        llvm::errs() << "CTU loaded AST file: "
-                     << ASTFileName << "\n";
+        llvm::errs() << "CTU loaded AST file: " << ASTSource << "\n";
       }
     } else {
       Unit = ASTCacheEntry->second.get();
     }
-    FunctionASTUnitMap[LookupName] = Unit;
+    // Fill the cache for the lookup name as well.
+    if (Unit)
```

Comment:
Can't we just `assert(Unit)`? If we have early returns and error handling on condition (!Unit) above then I guess we could.